### PR TITLE
Add test cases for function copyTexSubImage3D to detect feedback loop

### PR DESCRIPTION
**Pull request by JiangYizhou**
gl.INVALID_OPRATION is generated if the source and destination textures are the same.
Valid operation if the layer or level of soure and destination are different.

### CODE REVIEW

**Comment by JiangYizhou** [comment]
@Richard-Yunchao PTAL

**Comment by Richard-Yunchao** [comment]
LGTM. 

@zhenyao (and other owners), PTAL. Thanks a lot!

**Comment by Richard-Yunchao** [comment]
The first commit is OK. It can detect feedback loop for CopyTexSubImage3D. However, you have 2 commits now... 
BTW, I think it is better to write the code in conformance2/ file folder, instead of deqp. I prefer to keep WebGL deqp basically the same with native deqp. If we want to add some test cases, I think it is better to add the tests into conformance2/ file folder, instead of deqp/ file folder. WDYT? @zhenyao and @kenrussell ?

File: sdk/tests/deqp/functional/gles3/es3fNegativeTextureApiTests.js [hunk 1]
```diff
@@ -2455,6 +2455,145 @@ goog.scope(function() {
...
+                    if(pixels[x] != uint[x]) {
+                        this.testFailed("Get 0 if the textures is not initialized before copytexsubimage3d.");
+                        break;
+                    }
+                }
+            } else {
+                  this.testFailed("framebuffer not complete");
+            }
+
+            gl.bindTexture(gl.TEXRURE_3D, null);
```
Comment (kenrussell) [inline sdk/tests/deqp/functional/gles3/es3fNegativeTextureApiTests.js]: Critical typo: should be TEXTURE_3D.

**Comment by kenrussell** [comment]
Looking at the new tests, they don't have any significant dependency on the dEQP framework, so I agree with @Richard-Yunchao that it would be better to leave the dEQP port basically mirroring the original C++ code. Could you create a new file instead that incorporates these tests? Perhaps conformance2/textures/misc/copy-tex-image-webgl-specific.html? Thanks.

Also, please squash commits and re-upload with "git push origin [branchname] --force" per https://www.khronos.org/webgl/wiki/Using_Github_To_Contribute ? Thanks.

**Comment by JiangYizhou** [comment]
@Richard-Yunchao @kenrussell I have updated the code, PTAL.

**Comment by Richard-Yunchao** [comment]
LGTM. Could you have another look? @kenrussell , @zhenyao .

File: sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        gl.copyTexSubImage3D(gl.TEXTURE_3D, 0, 0, 0, layer, 0, 0, width, height);
+            wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "gl.INVALID_OPRATION is generated");
+    } else {
+        testFailed("framebuffer not complete");
+    }
+
+    gl.bindFramebuffer(gl.FRAMEBUFFER, null);
+    gl.deleteFramebuffer(fbo);
+    gl.deleteTexture(texture);
+    testPassed("gl.INVALID_OPRATION is generated if the source and destination textures are the same.");
```
Comment (kenrussell) [inline sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html]: OPRATION -> OPERATION

File: sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    var height = 2;
+    var depth = 2;
+    gl.texImage3D(gl.TEXTURE_3D, 0, gl.RGBA, width, height, depth, 0, gl.RGBA, gl.UNSIGNED_BYTE, uint8);
+
+    var fbo = gl.createFramebuffer();
+    gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
+    gl.framebufferTextureLayer(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, texture, 0, layer);
+    if (gl.checkFramebufferStatus(gl.FRAMEBUFFER) == gl.FRAMEBUFFER_COMPLETE) {
+        gl.copyTexSubImage3D(gl.TEXTURE_3D, 0, 0, 0, layer, 0, 0, width, height);
+            wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "gl.INVALID_OPRATION is generated");
```
Comment (kenrussell) [inline sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html]: OPRATION -> OPERATION

File: sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    gl.texImage3D(gl.TEXTURE_3D, 0, gl.RGBA, width, height, depth, 0, gl.RGBA, gl.UNSIGNED_BYTE, uint8);
+
+    var fbo = gl.createFramebuffer();
+    gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
+    gl.framebufferTextureLayer(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, texture, 0, layer1);
+    if (gl.checkFramebufferStatus(gl.FRAMEBUFFER) == gl.FRAMEBUFFER_COMPLETE) {
+        gl.copyTexSubImage3D(gl.TEXTURE_3D, 0, 0, 0, layer2, 0, 0, width, height);
+        wtu.glErrorShouldBe(gl, gl.NO_ERROR, "texImage3D should succeed.");
+    } else {
+        testFailed("framebuffer not complete");
```
Comment (kenrussell) [inline sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html]: Return early here.

File: sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
```
Comment (kenrussell) [inline sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html]: Return early.

File: sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    gl.texImage3D(gl.TEXTURE_3D, 0, gl.RGBA, width, height, depth, 0, gl.RGBA, gl.UNSIGNED_BYTE, uint8);
+
+    var fbo = gl.createFramebuffer();
+    gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
+    gl.framebufferTextureLayer(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, texture, 0, layer);
+    if (gl.checkFramebufferStatus(gl.FRAMEBUFFER) == gl.FRAMEBUFFER_COMPLETE) {
+        gl.copyTexSubImage3D(gl.TEXTURE_3D, 0, 0, 0, layer, 0, 0, width, height);
+            wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION, "gl.INVALID_OPRATION is generated");
+    } else {
+        testFailed("framebuffer not complete");
```
Comment (kenrussell) [inline sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html]: Return early.

**Comment by kenrussell** [comment]
Mostly looks good, thanks. Couple of minor points.

**Comment by JiangYizhou** [comment]
@kenrussell  Thank you, I have revised the code as your suggest. PTAL.

File: sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html (hunk 4 above)
Comment (Richard-Yunchao) [inline sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html]: The original testPassed() is not necessary, but I suggest Yizhou to keep the deinit code, no matter it fails or successes. 
So she deleted the original testPassed() after deinit. 
PTAL. @kenrussell and @zhenyao.

File: sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    var width = 2;
+    var height = 2;
+    var depth = 2;
+    gl.texImage3D(gl.TEXTURE_3D, 0, gl.RGBA, width, height, depth, 0, gl.RGBA, gl.UNSIGNED_BYTE, uint8);
+
+    var fbo = gl.createFramebuffer();
+    gl.bindFramebuffer(gl.FRAMEBUFFER, fbo);
+    gl.framebufferTextureLayer(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, texture, 0, layer1);
+    if (gl.checkFramebufferStatus(gl.FRAMEBUFFER) == gl.FRAMEBUFFER_COMPLETE) {
+        gl.copyTexSubImage3D(gl.TEXTURE_3D, 0, 0, 0, layer2, 0, 0, width, height);
```
Comment (zhenyao) [inline sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html]: I assume your plan was to test different level (from the function name), but you are testing different layers here.

File: sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html (hunk 4 above)
Comment (kenrussell) [inline sdk/tests/conformance2/textures/misc/copy-texture-image-webgl-specific.html]: Seems fine. Thanks for giving this more thought.

**Comment by kenrussell** [comment]
Thanks, this looks good to me once @zhenyao 's comments are addressed. @zhenyao , please merge when you're satisfied with the test.

**Comment by JiangYizhou** [comment]
Sorry to make a rookie mistake, I revised the code. PTAL.@zhenyao

**Comment by zhenyao** [comment]
lgtm

**Pull request merged**